test/upgrade-test.go

```diff
@@ -0,0 +1,590 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/Masterminds/semver"
+	"github.com/urfave/cli/v2"
+
+	"github.com/sourcegraph/run"
+)
+
+// This is the main entry point for a docker-compose upgrade test tool intended for use within CI/CD pipelines, but also usable by developers for quick smoke tests
+func main() {
+	app := &cli.App{
+		Name:  "upgrade-test",
+		Usage: "Upgrade test is an upgrade smoke test for Sourcegraph's docker-compose deployment type.",
+		Flags: []cli.Flag{
+			&cli.BoolFlag{
+				Name:    "verbose",
+				Aliases: []string{"v"},
+				Usage:   "Stream verbose output to stdout.",
+			},
+		},
+		Commands: []*cli.Command{
+			{
+				Name:    "standard",
+				Aliases: []string{"std"},
+				Usage:   "Runs a standard upgrade test. between specified versions \n\nExample:\n\nupgrade-test standard -vs 5.0.0,5.1.0",
+				Flags: []cli.Flag{
+					&cli.StringSliceFlag{
+						Name:     "versions",
+						Aliases:  []string{"vs"},
+						Usage:    "A sequence of versions to do standard upgrades through.",
+						Required: true,
+					},
+					&cli.BoolFlag{
+						Name:    "order-versions",
+						Aliases: []string{"ovs"},
+						Usage:   "Order versions in the sequence provided by the user, before commencing upgrades.",
+					},
+				},
+				Action: func(cCtx *cli.Context) error {
+					ctx := cCtx.Context
+					// get flags
+					verbose := cCtx.Bool("verbose")
+					order := cCtx.Bool("order-versions")
+					// Process user provided versions
+					userVersions := cCtx.StringSlice("versions")
+					var versions []*semver.Version
+					for _, v := range userVersions {
+						ver, err := semver.NewVersion(v)
+						if err != nil {
+							return fmt.Errorf("Invalid version: %s", v)
+						}
+
+						versions = append(versions, ver)
+					}
+
+					err := initTest(ctx, verbose)
+					if err != nil {
+						return fmt.Errorf(err.Error())
+					}
+					// Remove the /tmp deploy-sourcegraph-docker repo after test completion or error
+					defer func(ctx context.Context) error {
+						err := dockerClean(ctx, verbose)
+						if err != nil {
+							return fmt.Errorf(err.Error())
+						}
+						fmt.Println("Cleaning up deploy-sourcegraph-docker repo...")
+						err = run.Cmd(ctx, "rm", "-rf", "/tmp/deploy-sourcegraph-docker").Run().Wait()
+						if err != nil {
+							return fmt.Errorf("Failed to remove deploy-sourcegraph-docker repo: %w", err)
+						}
+						return nil
+					}(ctx)
```

Comment:
Alright in the latest commit I implemented a `testEnv` type and an `initTestEnv`. Definitely cleaner. I'm just printing any errors from the cleanup function now rather than passing the error back up.